### PR TITLE
*: support return prev deleted kv

### PR DESCRIPTION
**Pull request by xiang90**
I still need to add a few tests. But I would like to see the feedbacks.

Now the prev value is got from API layer. It is not super efficient. We can do it in mvcc layer. But I feel it is good enough for now.

What do you think /cc @heyitsanthony @gyuho @hongchaodeng @mitake

### CODE REVIEW

File: etcdctl/ctlv3/command/del_command.go [hunk 1]
```diff
@@ -34,6 +35,8 @@ func NewDelCommand() *cobra.Command {
 	}
 
 	cmd.Flags().BoolVar(&delPrefix, "prefix", false, "delete keys with matching prefix")
+	cmd.Flags().BoolVar(&delKVs, "deletedKVs", false, "get delted KVs")
```
Comment (gyuho) [inline etcdctl/ctlv3/command/del_command.go]: Something like `get-deleted` or `deleted-kvs` to follow other flag string patterns?

Comment (xiang90) [inline etcdctl/ctlv3/command/del_command.go]: oh, right.

**Comment by mitake** [comment]
@xiang90 the idea seems to be reasonable. How do you think about a case of write only permission? Current delete request just requires write permission. This change introduces read operation in the delete request. Maybe permission checking rule needs to be changed for the case of `GetDeletedKVs == true`?

File: etcdserver/etcdserverpb/rpc.proto [hunk 2]
```diff
@@ -404,12 +404,17 @@ message DeleteRangeRequest {
   // If range_end is not given, the range is defined to contain only the key argument.
   // If range_end is '\0', the range is all keys greater than or equal to the key argument.
   bytes range_end = 2;
+
+  // if getDeletedKVs is set, the deleted KVs will be returned.
+  bool getDeletedKVs = 3;
```
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: `get_kvs`?

File: etcdserver/etcdserverpb/rpc.proto [hunk 3]
```diff
@@ -404,12 +404,17 @@ message DeleteRangeRequest {
...
 }
 
 message DeleteRangeResponse {
   ResponseHeader header = 1;
   // deleted is the number of keys deleted by the delete range request.
   int64 deleted = 2;
+  // if deletedKVs is set in the request, the deleted KVs will be returned.
+  repeated mvccpb.KeyValue deletedKVs = 3;
```
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: just `kvs`?

File: clientv3/op.go [hunk 4]
```diff
@@ -258,6 +264,10 @@ func withTop(target SortTarget, order SortOrder) []OpOption {
 	return []OpOption{WithPrefix(), WithSort(target, order), WithLimit(1)}
 }
 
+func WithGetDeletedKVs() OpOption {
```
Comment (heyitsanthony) [inline clientv3/op.go]: `WithKVs`?

**Comment by heyitsanthony** [comment]
approach looks OK

File: clientv3/op.go (hunk 4 above)
Comment (hongchaodeng) [inline clientv3/op.go]: WithPrevKVs?

Comment (heyitsanthony) [inline clientv3/op.go]: It should probably match whatever it's called in DeleteResponse, which is looking very similar to RangeResponse.

Comment (xiang90) [inline clientv3/op.go]: I do not like prevKV. deletedKV is more obvious than prevKV if we want the prefix.

**Comment by xiang90** [comment]
@mitake I think it would require read permission.

**Comment by xiang90** [comment]
@heyitsanthony All fixed. PTAL.

@mitake Auth issue resolved. PTAL.

File: etcdctl/ctlv3/command/del_command.go [hunk 5]
```diff
@@ -34,6 +35,8 @@ func NewDelCommand() *cobra.Command {
 	}
 
 	cmd.Flags().BoolVar(&delPrefix, "prefix", false, "delete keys with matching prefix")
+	cmd.Flags().BoolVar(&delKVs, "preserve-kvs", false, "preserve delted KVs")
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/del_command.go]: s/delted/deleted
preserve and return deleted key-value pairs?

File: etcdserver/etcdserverpb/rpc.proto [hunk 6]
```diff
@@ -404,12 +404,19 @@ message DeleteRangeRequest {
   // If range_end is not given, the range is defined to contain only the key argument.
   // If range_end is '\0', the range is all keys greater than or equal to the key argument.
   bytes range_end = 2;
+
+  // If preserveKVs is set, the deleted KVs will be preserved.
```
Comment (heyitsanthony) [inline etcdserver/etcdserverpb/rpc.proto]: preserved for delete events?

**Comment by heyitsanthony** [comment]
lgtm after fixing final nits

**Comment by mitake** [comment]
@xiang90 lgtm, thanks

**Pull request merged**